src/atomic.rs

```diff
@@ -0,0 +1,448 @@
+        let ptr = self.inner.swap(obj.into_ptr(), order);
+        // SAFETY: `ptr` is a non-null, owned pointer to `ffi::PyObject` of type `T`
+        unsafe { Py::from_owned_ptr_unchecked(ptr) }
+    }
+
+    /// Stores `obj` in the [`AtomicPy`], returning the previous value.
+    ///
+    /// `swap` takes an [Ordering] argument which describes the memory ordering of this operation.
+    /// All ordering modes are possible. Note that using [Acquire](Ordering::Acquire) makes the
+    /// store part of this operation [Relaxed](Ordering::Relaxed), and using
+    /// [Release](Ordering::Release) makes the load part [Relaxed](Ordering::Relaxed).
+    #[inline]
+    pub fn swap<'py>(&self, obj: Bound<'py, T>, order: Ordering) -> Bound<'py, T> {
+        let py = obj.py();
+        self.swap_unbound(obj.unbind(), order).into_bound(py)
+    }
+
+    /// Stores `new` into the [`AtomicPy`] if its current value is `current` (Python `current is
+    /// self`)
+    ///
+    /// The return value is a [`Result`] indicating whether `new` was written and containing the
+    /// previous value.
+    ///
+    /// `compare_exchange` takes two [Ordering] arguments to describe the memory ordering of this
+    /// operation. `success` describes the required ordering for the read-modify-write operation
+    /// that takes place if the comparison with `current` succeeds. `failure` describes the required
+    /// ordering for the load operation that takes place when the comparison fails. Using
+    /// [Acquire](Ordering::Acquire) as success ordering makes the store part of this operation
+    /// [Relaxed](Ordering::Relaxed), and using [Release](Ordering::Release) makes the successful
+    /// load [Relaxed](Ordering::Relaxed). The failure ordering can only be
+    /// [SeqCst](Ordering::SeqCst), [Acquire](Ordering::Acquire) or [Relaxed](Ordering::Relaxed).
+    #[inline]
+    pub fn compare_exchange<'py>(
+        &self,
+        current: &Bound<'py, T>,
+        new: &Bound<'py, T>,
+        success: Ordering,
+        failure: Ordering,
+    ) -> Result<Bound<'py, T>, Bound<'py, T>> {
+        let py = current.py();
+        match self
+            .inner
+            .compare_exchange(current.as_ptr(), new.as_ptr(), success, failure)
+        {
+            // stored `new`, `ptr` is owned
+            Ok(ptr) => unsafe {
+                let _ = std::mem::ManuallyDrop::new(new.clone()); // only incref if `new` was successfully stored
+
+                // SAFETY: `ptr` is a non-null, owned pointer to `ffi::PyObject` of type `T`
+                Ok(Bound::from_owned_ptr_unchecked(py, ptr).cast_into_unchecked())
+            },
+            // did not store `new`, `ptr` is borrowed
+            Err(ptr) => unsafe {
+                // SAFETY: `ptr` is a non-null, borrowed pointer to `ffi::PyObject` of type `T`
+                Err(Borrowed::from_ptr_unchecked(py, ptr)
+                    .to_owned()
+                    .cast_into_unchecked())

```

Comment:
I wonder, is there a potential race here if another thread writes avalue to the `AtomicPy` dropping this `ptr` before the `.to_owned` call?

... is this potentially indicative of a more general problem with `AtomicPy`, e.g. does `load` have the same issue? 🤔

Reply:
Uff, I think you're right here. I guess this means we can only really provide the `swap` operation (and probably `take` for the option variant), because that does not touch the ref-count (at least if we want to keep this a thin wrapper). 

I could image providing `load` and `swap` if we take more control. 
- `load` would need to swap with nullptr, increment the ref-count, store back original
- `load` and `swap` spin briefly if we're in the middle of a `load`
- we can't expose an `Ordering` anymore

Not sure if that would have significant benefits left over a `Mutex<Py>`. 

`compare_exchange` looks pretty hopeless, I don't see a way to expose that.